data/trie/pruningBuffer.go

```diff
@@ -26,29 +26,19 @@ func (pb *pruningBuffer) add(rootHash []byte) {
 		return
 	}
 
-	pb.buffer[string(rootHash)] = struct{}{}
+	pb.buffer = append(pb.buffer, rootHash)
 	log.Trace("pruning buffer add", "rootHash", rootHash)
 }
 
-func (pb *pruningBuffer) remove(rootHash []byte) {
-	pb.mutOp.Lock()
-	defer pb.mutOp.Unlock()
-
-	delete(pb.buffer, string(rootHash))
-}
-
-func (pb *pruningBuffer) removeAll() map[string]struct{} {
+func (pb *pruningBuffer) removeAll() [][]byte {
 	pb.mutOp.Lock()
 	defer pb.mutOp.Unlock()
 
 	log.Trace("pruning buffer", "len", len(pb.buffer))
 
-	buffer := make(map[string]struct{})
-	for key := range pb.buffer {
-		buffer[key] = struct{}{}
-	}
-
-	pb.buffer = make(map[string]struct{})
+	buffer := make([][]byte, len(pb.buffer))
+	copy(buffer, pb.buffer)
+	pb.buffer = pb.buffer[:0]
```

Comment:
refactor this as this can lead to hidden bugs because you still can access the old values if you don't overwrite them
```
	sl := []int{1,2,3,4,5,6,7,8}
	sl = sl[:0]

	sl = append(sl, 7)
	fmt.Println(sl[:2])
```
will print [7,2]

Reply:
It's ok if you range the buffer, but also the capacity remains set to the old value. Please wait for a second opinion

Reply:
I think it could remain like this, as we range on this buffer, and the garbage remained beyond the buffer len should not be a problem. It is just an allocated memory which is useful for the future appends, if they will be. We use this pattern in many other parts of the project. You can search after [:0]. The alternative, is to replace this approach with make but this would cause to reallocate the memory to each append which would be done afterward.